python/graphstorm/model/edge_decoder.py

```diff
@@ -909,7 +909,58 @@ def predict_proba(self, g, h, e_h):
         return out
 
 ##################### Link Prediction Decoders #######################
-class LinkPredictNoParamDecoder(GSLayerNoParam):
+class LinkPredictionTestScoreMixin(abc.ABC):
```

Comment:
Why we add this Interface?

Reply:
After discussing with Jian it seems like all LP decoders need to implement this function, with the same signature. Adding this as an interface allows us to apply it to the two class hierarchies (`LinkPredictNoParamDecoder`, `LinkPredictLearnableDecoder`) as a mixin. Then we in `lp_gnn.py` we can have the assertion 

```
assert isinstance(decoder, LinkPredictionTestScoreInterface)
for pos_neg_tuple, neg_sample_type in loader:
    score: dict[tuple, tuple[th.Tensor, th.Tensor]] = \
        decoder.calc_test_scores(
            emb, pos_neg_tuple, neg_sample_type, device)
```

which guarantees that the decoder indeed should implement this function and informs the type inference of the function signature e.g.

<img width="686" alt="image" src="https://github.com/user-attachments/assets/31b3659e-fb80-4ed7-9204-7285b1ad51a9">

if we tried to pass a wrong argument type we get an error from the type-checker (`device` is of type `torch.device` but here we try to pass an `int`):

<img width="819" alt="image" src="https://github.com/user-attachments/assets/ea2dd2a2-fa92-4580-8e93-a24449e94276">